File: backend/internal/service/redis.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"strconv"
	"time"

	"github.com/go-redis/redis/v8"
	"github.com/zeni-x/backend/internal/store"
)

// RedisService Redis 服务
type RedisService struct {
	// No longer holds global config - connections are passed dynamically
}

// NewRedisService 创建 Redis 服务
func NewRedisService() *RedisService {
	return &RedisService{}
}

// connect 创建 Redis 连接
func (s *RedisService) connect(conn *store.Connection) (*redis.Client, error) {
	// Parse DB index from DatabaseName field (stored as string)
	dbIndex := 0
	if conn.DatabaseName != "" {
		if parsed, err := strconv.Atoi(conn.DatabaseName); err == nil {
			dbIndex = parsed
		}
	}

	client := redis.NewClient(&redis.Options{
		Addr:     fmt.Sprintf("%s:%d", conn.Host, conn.Port),
		Password: conn.Password,
		DB:       dbIndex,
	})

	ctx := context.Background()
	if err := client.Ping(ctx).Err(); err != nil {
		return nil, err
	}

	return client, nil
}

// RedisInfo Redis 服务器信息
type RedisInfo struct {
	Version          string `json:"version"`
	Host             string `json:"host"`
	Port             int    `json:"port"`
	Connected        bool   `json:"connected"`
	UsedMemory       string `json:"used_memory"`
	TotalKeys        int64  `json:"total_keys"`
	ConnectedClients int64  `json:"connected_clients"`
}

// GetInfo 获取 Redis 信息
func (s *RedisService) GetInfo(conn *store.Connection) (*RedisInfo, error) {
	client, err := s.connect(conn)
	if err != nil {
		return &RedisInfo{
			Host:      conn.Host,
			Port:      conn.Port,
			Connected: false,
		}, nil
	}
	defer client.Close()

	ctx := context.Background()

	// 获取 INFO
	info, err := client.Info(ctx).Result()
	if err != nil {
		return nil, err
	}

	// 解析版本信息
	var version, usedMemory string
	var connectedClients int64
	for _, line := range splitLines(info) {
		if len(line) > 14 && line[:14] == "redis_version:" {
			version = line[14:]
		}
		if len(line) > 17 && line[:17] == "used_memory_human:" {
			usedMemory = line[17:]
		}
		if len(line) > 18 && line[:18] == "connected_clients:" {
			fmt.Sscanf(line[18:], "%d", &connectedClients)
		}
	}

	// 获取 Key 数量
	dbSize, _ := client.DBSize(ctx).Result()

	return &RedisInfo{
		Version:          version,
		Host:             conn.Host,
		Port:             conn.Port,
		Connected:        true,
		UsedMemory:       usedMemory,
		TotalKeys:        dbSize,
		ConnectedClients: connectedClients,
	}, nil
}

// splitLines 分割行
func splitLines(s string) []string {
	var lines []string
	start := 0
	for i := 0; i < len(s); i++ {
		if s[i] == '\n' {
			line := s[start:i]
			if len(line) > 0 && line[len(line)-1] == '\r' {
				line = line[:len(line)-1]
			}
			lines = append(lines, line)
			start = i + 1
		}
	}
	if start < len(s) {
		lines = append(lines, s[start:])
	}
	return lines
}

// KeyInfo Key 信息
type KeyInfo struct {
	Key   string      `json:"key"`
	Type  string      `json:"type"`
	TTL   int64       `json:"ttl"`
	Value interface{} `json:"value,omitempty"`
}

// KeysResult Keys 列表结果
type KeysResult struct {
	Keys   []KeyInfo `json:"keys"`
	Cursor uint64    `json:"cursor"`
	Total  int64     `json:"total"`
}

// ListKeys 列出 Keys
func (s *RedisService) ListKeys(conn *store.Connection, pattern string, cursor uint64, count int64) (*KeysResult, error) {
	client, err := s.connect(conn)
	if err != nil {
		return nil, err
	}
	defer client.Close()

	ctx := context.Background()

	if pattern == "" {
		pattern = "*"
	}

	// 使用 SCAN 迭代
	keys, nextCursor, err := client.Scan(ctx, cursor, pattern, count).Result()
	if err != nil {
		return nil, err
	}

	// 获取每个 Key 的类型和 TTL
	var keyInfos []KeyInfo
	for _, key := range keys {
		keyType, _ := client.Type(ctx, key).Result()
		ttl, _ := client.TTL(ctx, key).Result()

		keyInfos = append(keyInfos, KeyInfo{
			Key:  key,
			Type: keyType,
			TTL:  int64(ttl.Seconds()),
		})
	}

	// 获取总数
	total, _ := client.DBSize(ctx).Result()

	return &KeysResult{
		Keys:   keyInfos,
		Cursor: nextCursor,
		Total:  total,
	}, nil
}

// GetKey 获取 Key 详情
func (s *RedisService) GetKey(conn *store.Connection, key string) (*KeyInfo, error) {
	client, err := s.connect(conn)
	if err != nil {
		return nil, err
	}
	defer client.Close()

	ctx := context.Background()

	// 获取类型
	keyType, err := client.Type(ctx, key).Result()
	if err != nil {
		return nil, err
	}

	if keyType == "none" {
		return nil, fmt.Errorf("key not found: %s", key)
	}

	// 获取 TTL
	ttl, _ := client.TTL(ctx, key).Result()

	info := &KeyInfo{
		Key:  key,
		Type: keyType,
		TTL:  int64(ttl.Seconds()),
	}

	// 根据类型获取值
	switch keyType {
	case "string":
		val, err := client.Get(ctx, key).Result()
		if err != nil {
			return nil, err
		}
		info.Value = val

	case "hash":
		val, err := client.HGetAll(ctx, key).Result()
		if err != nil {
			return nil, err
		}
		info.Value = val

	case "list":
		val, err := client.LRange(ctx, key, 0, -1).Result()
		if err != nil {
			return nil, err
		}
		info.Value = val

	case "set":
		val, err := client.SMembers(ctx, key).Result()
		if err != nil {
			return nil, err
		}
		info.Value = val

	case "zset":
		val, err := client.ZRangeWithScores(ctx, key, 0, -1).Result()
		if err != nil {
			return nil, err
		}
		info.Value = val
	}

	return info, nil
}

// SetKeyRequest 设置 Key 请求
type SetKeyRequest struct {
	Key   string      `json:"key"`
	Type  string      `json:"type"`
	Value interface{} `json:"value"`
	TTL   int64       `json:"ttl,omitempty"` // 秒，-1 表示不过期
}

// SetKey 设置 Key
func (s *RedisService) SetKey(conn *store.Connection, req *SetKeyRequest) error {
	client, err := s.connect(conn)
	if err != nil {
		return err
	}
	defer client.Close()

	ctx := context.Background()

	var expiration time.Duration
	if req.TTL > 0 {
		expiration = time.Duration(req.TTL) * time.Second
	}

	switch req.Type {
	case "string":
		val, ok := req.Value.(string)
		if !ok {
			return fmt.Errorf("invalid value type for string")
		}
		return client.Set(ctx, req.Key, val, expiration).Err()

	case "hash":
		val, ok := req.Value.(map[string]interface{})
		if !ok {
			return fmt.Errorf("invalid value type for hash")
		}
		if err := client.Del(ctx, req.Key).Err(); err != nil {
			return err
		}
		for k, v := range val {
			if err := client.HSet(ctx, req.Key, k, v).Err(); err != nil {
				return err
			}
		}
		if expiration > 0 {
			client.Expire(ctx, req.Key, expiration)
		}
		return nil

	case "list":
		val, ok := req.Value.([]interface{})
		if !ok {
			return fmt.Errorf("invalid value type for list")
		}
		if err := client.Del(ctx, req.Key).Err(); err != nil {
			return err
		}
		for _, v := range val {
			if err := client.RPush(ctx, req.Key, v).Err(); err != nil {
				return err
			}
		}
		if expiration > 0 {
			client.Expire(ctx, req.Key, expiration)
		}
		return nil

	case "set":
		val, ok := req.Value.([]interface{})
		if !ok {
			return fmt.Errorf("invalid value type for set")
		}
		if err := client.Del(ctx, req.Key).Err(); err != nil {
			return err
		}
		for _, v := range val {
			if err := client.SAdd(ctx, req.Key, v).Err(); err != nil {
				return err
			}
		}
		if expiration > 0 {
			client.Expire(ctx, req.Key, expiration)
		}
		return nil

	case "zset":
		val, ok := req.Value.([]interface{})
		if !ok {
			return fmt.Errorf("invalid value type for zset")
		}
		if err := client.Del(ctx, req.Key).Err(); err != nil {
			return err
		}
		for _, item := range val {
			m, ok := item.(map[string]interface{})
			if !ok {
				continue
			}
			member, _ := m["member"].(string)
			score, _ := m["score"].(float64)
			if err := client.ZAdd(ctx, req.Key, &redis.Z{Score: score, Member: member}).Err(); err != nil {
				return err
			}
		}
		if expiration > 0 {
			client.Expire(ctx, req.Key, expiration)
		}
		return nil

	default:
		return fmt.Errorf("unsupported type: %s", req.Type)
	}
}

// DeleteKey 删除 Key
func (s *RedisService) DeleteKey(conn *store.Connection, key string) error {
	client, err := s.connect(conn)
	if err != nil {
		return err
	}
	defer client.Close()

	return client.Del(context.Background(), key).Err()
}

// SetTTL 设置 TTL
func (s *RedisService) SetTTL(conn *store.Connection, key string, ttl int64) error {
	client, err := s.connect(conn)
	if err != nil {
		return err
	}
	defer client.Close()

	ctx := context.Background()

	if ttl < 0 {
		// 移除过期时间
		return client.Persist(ctx, key).Err()
	}

	return client.Expire(ctx, key, time.Duration(ttl)*time.Second).Err()
}

// ExportData 导出数据
type ExportData struct {
	Keys []KeyInfo `json:"keys"`
}

// Export 导出数据
func (s *RedisService) Export(conn *store.Connection, keys []string) (*ExportData, error) {
	var keyInfos []KeyInfo

	for _, key := range keys {
		info, err := s.GetKey(conn, key)
		if err != nil {
			continue
		}
		keyInfos = append(keyInfos, *info)
	}

	return &ExportData{Keys: keyInfos}, nil
}

// Import 导入数据
func (s *RedisService) Import(conn *store.Connection, data *ExportData) error {
	for _, keyInfo := range data.Keys {
		req := &SetKeyRequest{
			Key:   keyInfo.Key,
			Type:  keyInfo.Type,
			Value: keyInfo.Value,
			TTL:   keyInfo.TTL,
		}
		if err := s.SetKey(conn, req); err != nil {
			return err
		}
	}
	return nil
}

// ExportJSON 导出为 JSON 字符串
func (s *RedisService) ExportJSON(conn *store.Connection, keys []string) (string, error) {
	data, err := s.Export(conn, keys)
	if err != nil {
		return "", err
	}

	b, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return "", err
	}

	return string(b), nil
}

// ImportJSON 从 JSON 导入
func (s *RedisService) ImportJSON(conn *store.Connection, jsonStr string) error {
	var data ExportData
	if err := json.Unmarshal([]byte(jsonStr), &data); err != nil {
		return err
	}

	return s.Import(conn, &data)
}

// TestConnection 测试连接是否有效
func (s *RedisService) TestConnection(conn *store.Connection) error {
	client, err := s.connect(conn)
	if err != nil {
		return err
	}
	defer client.Close()
	return nil
}

File: src-tauri/src/services/redis.rs
//! Redis database operations service
//!
//! This service handles all Redis-related operations including:
//! - Server info retrieval
//! - Key management (CRUD operations)
//! - TTL management
//! - Export/Import

use redis::aio::ConnectionManager;
use redis::{AsyncCommands, Client, Value as RedisValue};
use serde_json::Value as JsonValue;

use crate::db::models::{
    Connection, RedisExportData, RedisKeyInfo, RedisKeyListResponse, RedisKeyValue,
    RedisServerInfo, SetKeyRequest,
};
use crate::error::{AppError, AppResult};

/// Redis service for database operations
pub struct RedisService {
    manager: ConnectionManager,
    connection: Connection,
}

impl RedisService {
    /// Create a new Redis service by connecting to the server
    pub async fn connect(conn: &Connection) -> AppResult<Self> {
        let password = conn.password.as_deref().unwrap_or("");
        let db = conn.database_name.as_deref().unwrap_or("0");

        // For K8s connections, use forward_local_port if available (port forwarding active)
        // Otherwise fall back to the original port
        let effective_port = conn
            .forward_local_port
            .filter(|&p| p > 0)
            .unwrap_or(conn.port);

        log::info!(
            "RedisService::connect - connection_id: {:?}, password provided: {}, password length: {}, host: {}, port: {} (forward_local_port: {:?})",
            conn.id,
            !password.is_empty(),
            password.len(),
            conn.host,
            effective_port,
            conn.forward_local_port
        );

        let url = if password.is_empty() {
            format!("redis://{}:{}/{}", conn.host, effective_port, db)
        } else {
            format!("redis://:{}@{}:{}/{}", password, conn.host, effective_port, db)
        };

        let client = Client::open(url).map_err(|e| AppError::Connection(e.to_string()))?;

        let manager = ConnectionManager::new(client)
            .await
            .map_err(|e| AppError::Connection(e.to_string()))?;

        Ok(Self {
            manager,
            connection: conn.clone(),
        })
    }

    /// Get Redis server info
    pub async fn get_info(&mut self) -> AppResult<RedisServerInfo> {
        let info: String = redis::cmd("INFO")
            .query_async(&mut self.manager)
            .await
            .map_err(|e| AppError::Database(e.to_string()))?;

        // Parse INFO response
        let mut version = String::new();
        let mut used_memory = None;
        let mut connected_clients = None;
        let mut uptime_seconds = None;

        for line in info.lines() {
            if let Some(v) = line.strip_prefix("redis_version:") {
                version = v.to_string();
            } else if let Some(v) = line.strip_prefix("used_memory_human:") {
                used_memory = Some(v.to_string());
            } else if let Some(v) = line.strip_prefix("connected_clients:") {
                connected_clients = v.parse().ok();
            } else if let Some(v) = line.strip_prefix("uptime_in_seconds:") {
                uptime_seconds = v.parse().ok();
            }
        }

        // Get database count
        let db_count = self.get_db_count().await.unwrap_or(16);

        Ok(RedisServerInfo {
            version,
            host: self.connection.host.clone(),
            port: self.connection.port,
            connected: true,
            used_memory,
            connected_clients,
            uptime_seconds,
            db_count,
        })
    }

    /// Get number of databases
    async fn get_db_count(&mut self) -> AppResult<i64> {
        let config: Vec<String> = redis::cmd("CONFIG")
            .arg("GET")
            .arg("databases")
            .query_async(&mut self.manager)
            .await
            .map_err(|e| AppError::Database(e.to_string()))?;

        if config.len() >= 2 {
            Ok(config[1].parse().unwrap_or(16))
        } else {
            Ok(16)
        }
    }

    /// List keys with pattern and cursor-based pagination
    pub async fn list_keys(
        &mut self,
        pattern: &str,
        cursor: u64,
        count: u64,
    ) -> AppResult<RedisKeyListResponse> {
        let pattern = if pattern.is_empty() { "*" } else { pattern };

        let (new_cursor, keys): (u64, Vec<String>) = redis::cmd("SCAN")
            .arg(cursor)
            .arg("MATCH")
            .arg(pattern)
            .arg("COUNT")
            .arg(count)
            .query_async(&mut self.manager)
            .await
            .map_err(|e| AppError::Database(e.to_string()))?;

        let mut key_infos = Vec::new();

        for key in &keys {
            let key_type = self.get_key_type(key).await.unwrap_or_else(|_| "unknown".to_string());
            let ttl = self.get_ttl(key).await.unwrap_or(-1);

            key_infos.push(RedisKeyInfo {
                key: key.clone(),
                key_type,
                ttl,
                value: None,
                size: None,
            });
        }

        Ok(RedisKeyListResponse {
            keys: key_infos,
            cursor: new_cursor,
            has_more: new_cursor != 0,
        })
    }

    /// Get key type
    async fn get_key_type(&mut self, key: &str) -> AppResult<String> {
        let key_type: String = redis::cmd("TYPE")
            .arg(key)
            .query_async(&mut self.manager)
            .await
            .map_err(|e| AppError::Database(e.to_string()))?;
        Ok(key_type)
    }

    /// Get TTL for a key
    async fn get_ttl(&mut self, key: &str) -> AppResult<i64> {
        let ttl: i64 = self
            .manager
            .ttl(key)
            .await
            .map_err(|e| AppError::Database(e.to_string()))?;
        Ok(ttl)
    }

    /// Get key value
    pub async fn get_key(&mut self, key: &str) -> AppResult<RedisKeyValue> {
        let key_type = self.get_key_type(key).await?;
        let ttl = self.get_ttl(key).await.unwrap_or(-1);

        let value = match key_type.as_str() {
            "string" => {
                let v: String = self
                    .manager
                    .get(key)
                    .await
                    .map_err(|e| AppError::Database(e.to_string()))?;
                JsonValue::String(v)
            }
            "list" => {
                let v: Vec<String> = self
                    .manager
                    .lrange(key, 0, -1)
                    .await
                    .map_err(|e| AppError::Database(e.to_string()))?;
                JsonValue::Array(v.into_iter().map(JsonValue::String).collect())
            }
            "set" => {
                let v: Vec<String> = self
                    .manager
                    .smembers(key)
                    .await
                    .map_err(|e| AppError::Database(e.to_string()))?;
                JsonValue::Array(v.into_iter().map(JsonValue::String).collect())
            }
            "zset" => {
                let v: Vec<(String, f64)> = self
                    .manager
                    .zrange_withscores(key, 0, -1)
                    .await
                    .map_err(|e| AppError::Database(e.to_string()))?;
                let items: Vec<JsonValue> = v
                    .into_iter()
                    .map(|(member, score)| {
                        serde_json::json!({
                            "member": member,
                            "score": score
                        })
                    })
                    .collect();
                JsonValue::Array(items)
            }
            "hash" => {
                let v: Vec<(String, String)> = self
                    .manager
                    .hgetall(key)
                    .await
                    .map_err(|e| AppError::Database(e.to_string()))?;
                let obj: serde_json::Map<String, JsonValue> = v
                    .into_iter()
                    .map(|(k, v)| (k, JsonValue::String(v)))
                    .collect();
                JsonValue::Object(obj)
            }
            _ => JsonValue::Null,
        };

        Ok(RedisKeyValue {
            key: key.to_string(),
            key_type,
            ttl,
            value,
        })
    }

    /// Set a key value
    pub async fn set_key(&mut self, req: &SetKeyRequest) -> AppResult<()> {
        match req.key_type.as_str() {
            "string" => {
                let value = match &req.value {
                    JsonValue::String(s) => s.clone(),
                    _ => req.value.to_string(),
                };

                if let Some(ttl) = req.ttl {
                    if ttl > 0 {
                        let _: () = self
                            .manager
                            .set_ex(&req.key, value, ttl as u64)
                            .await
                            .map_err(|e| AppError::Database(e.to_string()))?;
                    } else {
                        let _: () = self
                            .manager
                            .set(&req.key, value)
                            .await
                            .map_err(|e| AppError::Database(e.to_string()))?;
                    }
                } else {
                    let _: () = self
                        .manager
                        .set(&req.key, value)
                        .await
                        .map_err(|e| AppError::Database(e.to_string()))?;
                }
            }
            "list" => {
                // Delete existing key first
                let _: () = self
                    .manager
                    .del(&req.key)
                    .await
                    .map_err(|e| AppError::Database(e.to_string()))?;

                if let JsonValue::Array(arr) = &req.value {
                    let values: Vec<String> = arr
                        .iter()
                        .map(|v| match v {
                            JsonValue::String(s) => s.clone(),
                            _ => v.to_string(),
                        })
                        .collect();

                    if !values.is_empty() {
                        let _: () = self
                            .manager
                            .rpush(&req.key, values)
                            .await
                            .map_err(|e| AppError::Database(e.to_string()))?;
                    }
                }

                if let Some(ttl) = req.ttl {
                    if ttl > 0 {
                        let _: () = self
                            .manager
                            .expire(&req.key, ttl)
                            .await
                            .map_err(|e| AppError::Database(e.to_string()))?;
                    }
                }
            }
            "set" => {
                // Delete existing key first
                let _: () = self
                    .manager
                    .del(&req.key)
                    .await
                    .map_err(|e| AppError::Database(e.to_string()))?;

                if let JsonValue::Array(arr) = &req.value {
                    let values: Vec<String> = arr
                        .iter()
                        .map(|v| match v {
                            JsonValue::String(s) => s.clone(),
                            _ => v.to_string(),
                        })
                        .collect();

                    if !values.is_empty() {
                        let _: () = self
                            .manager
                            .sadd(&req.key, values)
                            .await
                            .map_err(|e| AppError::Database(e.to_string()))?;
                    }
                }

                if let Some(ttl) = req.ttl {
                    if ttl > 0 {
                        let _: () = self
                            .manager
                            .expire(&req.key, ttl)
                            .await
                            .map_err(|e| AppError::Database(e.to_string()))?;
                    }
                }
            }
            "zset" => {
                // Delete existing key first
                let _: () = self
                    .manager
                    .del(&req.key)
                    .await
                    .map_err(|e| AppError::Database(e.to_string()))?;

                if let JsonValue::Array(arr) = &req.value {
                    for item in arr {
                        if let (Some(member), Some(score)) = (
                            item.get("member").and_then(|v| v.as_str()),
                            item.get("score").and_then(|v| v.as_f64()),
                        ) {
                            let _: () = self
                                .manager
                                .zadd(&req.key, member, score)
                                .await
                                .map_err(|e| AppError::Database(e.to_string()))?;
                        }
                    }
                }

                if let Some(ttl) = req.ttl {
                    if ttl > 0 {
                        let _: () = self
                            .manager
                            .expire(&req.key, ttl)
                            .await
                            .map_err(|e| AppError::Database(e.to_string()))?;
                    }
                }
            }
            "hash" => {
                // Delete existing key first
                let _: () = self
                    .manager
                    .del(&req.key)
                    .await
                    .map_err(|e| AppError::Database(e.to_string()))?;

                if let JsonValue::Object(obj) = &req.value {
                    let fields: Vec<(String, String)> = obj
                        .iter()
                        .map(|(k, v)| {
                            let val = match v {
                                JsonValue::String(s) => s.clone(),
                                _ => v.to_string(),
                            };
                            (k.clone(), val)
                        })
                        .collect();

                    if !fields.is_empty() {
                        let _: () = self
                            .manager
                            .hset_multiple(&req.key, &fields)
                            .await
                            .map_err(|e| AppError::Database(e.to_string()))?;
                    }
                }

                if let Some(ttl) = req.ttl {
                    if ttl > 0 {
                        let _: () = self
                            .manager
                            .expire(&req.key, ttl)
                            .await
                            .map_err(|e| AppError::Database(e.to_string()))?;
                    }
                }
            }
            _ => {
                return Err(AppError::Validation(format!(
                    "Unsupported key type: {}",
                    req.key_type
                )));
            }
        }

        Ok(())
    }

    /// Update a key (alias for set_key, but requires key to exist)
    pub async fn update_key(&mut self, key: &str, req: &SetKeyRequest) -> AppResult<()> {
        // Check if key exists
        let exists: bool = self
            .manager
            .exists(key)
            .await
            .map_err(|e| AppError::Database(e.to_string()))?;

        if !exists {
            return Err(AppError::NotFound(format!("Key not found: {}", key)));
        }

        self.set_key(req).await
    }

    /// Delete a key
    pub async fn delete_key(&mut self, key: &str) -> AppResult<()> {
        let _: () = self
            .manager
            .del(key)
            .await
            .map_err(|e| AppError::Database(e.to_string()))?;
        Ok(())
    }

    /// Set TTL for a key
    pub async fn set_ttl(&mut self, key: &str, ttl: i64) -> AppResult<()> {
        if ttl > 0 {
            let _: () = self
                .manager
                .expire(key, ttl)
                .await
                .map_err(|e| AppError::Database(e.to_string()))?;
        } else {
            // Remove TTL (persist)
            let _: () = self
                .manager
                .persist(key)
                .await
                .map_err(|e| AppError::Database(e.to_string()))?;
        }
        Ok(())
    }

    /// Export keys
    pub async fn export_keys(&mut self, keys: &[String]) -> AppResult<RedisExportData> {
        let mut exported = Vec::new();

        for key in keys {
            match self.get_key(key).await {
                Ok(kv) => exported.push(kv),
                Err(e) => {
                    log::warn!("Failed to export key {}: {}", key, e);
                }
            }
        }

        Ok(RedisExportData { keys: exported })
    }

    /// Import keys
    pub async fn import_keys(&mut self, data: &RedisExportData) -> AppResult<i32> {
        let mut imported = 0;

        for kv in &data.keys {
            let req = SetKeyRequest {
                key: kv.key.clone(),
                key_type: kv.key_type.clone(),
                value: kv.value.clone(),
                ttl: if kv.ttl > 0 { Some(kv.ttl) } else { None },
            };

            match self.set_key(&req).await {
                Ok(_) => imported += 1,
                Err(e) => {
                    log::warn!("Failed to import key {}: {}", kv.key, e);
                }
            }
        }

        Ok(imported)
    }
}
